buffer/src/main/java/io/netty/buffer/SizeClasses.java

```diff
@@ -0,0 +1,403 @@
+ *   ( 3,     4,        4,         3,       no,             yes,        4)
+ * <p>
+ *   ( 4,     6,        4,         1,       no,             yes,        4)
+ *   ( 5,     6,        4,         2,       no,             yes,        4)
+ *   ( 6,     6,        4,         3,       no,             yes,        4)
+ *   ( 7,     6,        4,         4,       no,             yes,        4)
+ * <p>
+ *   ( 8,     7,        5,         1,       no,             yes,        5)
+ *   ( 9,     7,        5,         2,       no,             yes,        5)
+ *   ( 10,    7,        5,         3,       no,             yes,        5)
+ *   ( 11,    7,        5,         4,       no,             yes,        5)
+ *   ...
+ *   ...
+ *   ( 72,    23,       21,        1,       yes,            no,        no)
+ *   ( 73,    23,       21,        2,       yes,            no,        no)
+ *   ( 74,    23,       21,        3,       yes,            no,        no)
+ *   ( 75,    23,       21,        4,       yes,            no,        no)
+ * <p>
+ *   ( 76,    24,       22,        1,       yes,            no,        no)
+ */
+public abstract class SizeClasses implements SizeClassesMetric {
+
+    static final int LOG2_QUANTUM = 4;
+
+    private static final int LOG2_SIZE_CLASS_GROUP = 2;
+    private static final int LOG2_MAX_LOOKUP_SIZE = 12;
+
+    private static final int INDEX_IDX = 0;
+    private static final int LOG2GROUP_IDX = 1;
+    private static final int LOG2DELTA_IDX = 2;
+    private static final int NDELTA_IDX = 3;
+    private static final int PAGESIZE_IDX = 4;
+    private static final int SUBPAGE_IDX = 5;
+    private static final int LOG2_DELTA_LOOKUP_IDX = 6;
+
+    private static final byte no = 0, yes = 1;
+
+    protected SizeClasses(int pageSize, int pageShifts, int chunkSize, int directMemoryCacheAlignment) {
+        this.pageSize = pageSize;
+        this.pageShifts = pageShifts;
+        this.chunkSize = chunkSize;
+        this.directMemoryCacheAlignment = directMemoryCacheAlignment;
+
+        int group = log2(chunkSize) + 1 - LOG2_QUANTUM;
+        sizeClasses = new short[group << LOG2_SIZE_CLASS_GROUP][7];
+
+        //generate size classes
+        sizeClasses();
+        //generate lookup table
+        idx2SizeTab();
+        size2idxTab();
+    }
+
+    protected final int pageSize;
+    protected final int pageShifts;
+    protected final int chunkSize;
+    protected final int directMemoryCacheAlignment;
+
+    int nSizes, nSubpages, nPSizes;
+
+    int smallMaxSizeIdx;
+
+    private int lookupMaxSize;
+
+    private final short[][] sizeClasses;
+
+    private int[] pageIdx2sizeTab;
+
+    // lookup table for sizeIdx <= smallMaxSizeIdx
+    private int[] sizeIdx2sizeTab;
+
+    // lookup table used for size <= lookupMaxclass
+    // spacing is 1 << LOG2_QUANTUM, so the size of array is lookupMaxclass >> LOG2_QUANTUM
+    private int[] size2idxTab;
+
+    private void sizeClasses() {
```

Comment:
We should inline this method into the constructor to be able to define `nSizes` final